veps/sig-compute/hypervisor-feature-discovery.md

```diff
@@ -0,0 +1,372 @@
+
+2. **Mark a feature as experimental**: Register as `Experimental` with a feature gate requirement
+   - Example: CPU hotplug while still under development
+   - Example: New network multiqueue features
+
+3. **Mark a feature as deprecated**: Register as `Deprecated` to warn users
+   - Feature still works but users should migrate away
+   - Helps communicate deprecation timeline
+
+4. **Enable test filtering**: Register to control which tests run on which platforms
+   - Tests for unsupported features automatically skip
+   - Reduces noise in CI for platform-specific features
+
+**What not to register**: Features where support is unknown or still being discovered. If a feature hasn't shown platform-specific issues and doesn't require explicit blocking, leave it unregistered. Let natural test failures and runtime errors reveal compatibility issues, then register capabilities as constraints are discovered. This keeps the registry focused on known validation needs rather than attempting to preemptively catalog all features.
+
+### Registration Examples
+
+Here are examples showing how hypervisors register capability constraints:
+
+```go
+// Usage in init() - hypervisor implementations register support:
+func init() {
+    // Register base KVM constraints (applies to all KVM archs unless overridden)
+    Register("kvm", "").  // empty arch = hypervisor-wide default
+      Experimental(CapCPUHotplug)
+    
+    // Architecture-specific constraints for KVM
+    Register("kvm", "arm64").
+      Cap(CapGraphicsVGA).
+        Unsupported().
+        WithMessage("VGA graphics not supported on ARM64 architecture").
+        WithDocLink("https://kubevirt.io/user-guide/graphics#compatibility").
+        Done()
+    
+    // VGA is unregistered on amd64/s390x - implicitly allowed
+}
+```
+
+### Resolution Strategy
+
+The registry resolves capabilities using a layered fallback strategy, checking from most specific to least specific:
+
+```go
+// Query with automatic fallback: hypervisor/arch -> hypervisor -> base
+func Get(hypervisor, arch string, cap CapabilityKey) Capability {
+    // Try exact match: kvm/amd64
+    if c, ok := lookup(hypervisor+"/"+arch, cap); ok { return c }
+    
+    // Try hypervisor-wide: kvm
+    if c, ok := lookup(hypervisor, cap); ok { return c }
+    
+    // Not registered - return zero value (caller checks cap.Level)
+    return Capability{}
+}
+```
+
+Resolution precedence (most specific wins):
+1. **Hypervisor+Architecture** (e.g., `kvm/amd64`) - most specific
+2. **Hypervisor-wide** (e.g., `kvm`) - applies to all architectures unless overridden
+3. **Not registered** - returns zero-value `Capability{}` (where `Level` is 0, which equals `Unregistered`)
+
+### Integration Points
+
+#### Validation Webhooks
+
+The validating webhook integrates with the capability registry by:
+
+1. Mapping VMI spec features to capability keys (e.g., `video.type: vga` → `graphics.vga`)
```

Comment:
Do we expect capabilities to map 1:1 with VMI spec fields? And if so, could we name capabilities so they could be inferred from API fields?
I find the matrix mechanism to be quite elegant, but having to maintain mapping functions is unfortunate...